File: services/api/app/schemas/offline_video.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.offline_video import OfflineVideoLabel, OfflineVideoOrigin


class OfflineVideoRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    relative_path: str
    file_name: str
    display_name: str
    dataset_name: str | None
    origin: OfflineVideoOrigin
    label: OfflineVideoLabel
    media_type: str
    size_bytes: int
    source_url: str | None
    license_note: str | None
    is_available: bool
    file_modified_at: datetime
    last_scanned_at: datetime
    created_at: datetime
    updated_at: datetime
    requires_transcoding: bool


class OfflineVideoUpdate(BaseModel):
    display_name: str | None = Field(default=None, min_length=1, max_length=200)
    dataset_name: str | None = Field(default=None, max_length=120)
    origin: OfflineVideoOrigin | None = None
    label: OfflineVideoLabel | None = None
    source_url: str | None = Field(default=None, max_length=500)
    license_note: str | None = Field(default=None, max_length=1000)

    @field_validator("display_name")
    @classmethod
    def normalize_display_name(cls, value: str | None) -> str:
        if value is None or not value.strip():
            raise ValueError("display_name cannot be empty")
        return value.strip()

    @field_validator("dataset_name", "source_url", "license_note")
    @classmethod
    def normalize_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None


class OfflineVideoLibraryStatus(BaseModel):
    root_hint: str
    supported_extensions: list[str]
    total_count: int
    available_count: int
    labeled_count: int
    dataset_count: int
    last_scanned_at: datetime | None
    inference_enabled: bool = False
    transcoding_enabled: bool = True


class OfflineVideoScanResponse(BaseModel):
    created: int
    updated: int
    missing: int
    total: int
    scanned_at: datetime


class OfflineVideoPlaybackTicket(BaseModel):
    url: str
    expires_at: datetime
    transcoded: bool

File: apps/web/src/types/offline-video.ts
export type OfflineVideoOrigin = 'public_dataset' | 'self_recorded' | 'synthetic' | 'other'
export type OfflineVideoLabel = 'fall' | 'adl' | 'near_fall' | 'unknown'

export interface OfflineVideo {
  id: number
  relative_path: string
  file_name: string
  display_name: string
  dataset_name: string | null
  origin: OfflineVideoOrigin
  label: OfflineVideoLabel
  media_type: string
  size_bytes: number
  source_url: string | null
  license_note: string | null
  is_available: boolean
  file_modified_at: string
  last_scanned_at: string
  created_at: string
  updated_at: string
  requires_transcoding: boolean
}

export interface OfflineVideoLibraryStatus {
  root_hint: string
  supported_extensions: string[]
  total_count: number
  available_count: number
  labeled_count: number
  dataset_count: number
  last_scanned_at: string | null
  inference_enabled: boolean
  transcoding_enabled: boolean
}

export interface OfflineVideoScanResult {
  created: number
  updated: number
  missing: number
  total: number
  scanned_at: string
}

export interface OfflineVideoUpdate {
  display_name?: string | null
  dataset_name?: string | null
  origin?: OfflineVideoOrigin
  label?: OfflineVideoLabel
  source_url?: string | null
  license_note?: string | null
}

export interface OfflineVideoPlaybackTicket {
  url: string
  expires_at: string
  transcoded: boolean
}
